idea/src/org/jetbrains/kotlin/idea/intentions/ConvertFilteringFunctionWithDemorgansLawIntention.kt

```diff
@@ -0,0 +1,150 @@
+        if (element.getResolvedCall(context)?.resultingDescriptor?.fqNameOrNull() !in fqNames) return null
+
+        val lambda = element.lambda() ?: return null
+        val lambdaBody = lambda.bodyExpression ?: return null
+        if (lambdaBody.anyDescendantOfType<KtReturnExpression>()) return null
+        if (lambdaBody.statements.lastOrNull()?.getType(context)?.isBoolean() != true) return null
+
+        return callee.textRange
+    }
+
+    override fun applyTo(element: KtCallExpression, editor: Editor?) {
+        val lambda = element.lambda() ?: return
+        val lastExpression = lambda.bodyExpression?.statements?.lastOrNull() ?: return
+        val callOrQualified = element.getQualifiedExpressionForSelector() ?: element
+        val psiFactory = KtPsiFactory(element)
+
+        if (negatePredicate) {
+            val exclPrefixExpression = lastExpression.asExclPrefixExpression()
+            if (exclPrefixExpression == null) {
+                val replaced = lastExpression.replaced(psiFactory.createExpressionByPattern("!($0)", lastExpression)) as KtPrefixExpression
+                replaced.baseExpression.removeUnnecessaryParentheses()
+                when (val baseExpression = replaced.baseExpression?.deparenthesize()) {
+                    is KtBinaryExpression -> {
+                        val operationToken = baseExpression.operationToken
+                        if (operationToken == KtTokens.ANDAND || operationToken == KtTokens.OROR) {
+                            ConvertBinaryExpressionWithDemorgansLawIntention.convertIfPossible(baseExpression)
+                        } else {
+                            SimplifyNegatedBinaryExpressionInspection.simplifyNegatedBinaryExpressionIfNeeded(replaced)
+                        }
+                    }
+                    is KtQualifiedExpression -> {
+                        baseExpression.invertSelectorFunction()?.let { replaced.replace(it) }
+                    }
+                }
+            } else {
+                val replaced = exclPrefixExpression.baseExpression?.let { lastExpression.replaced(it) }
+                replaced.removeUnnecessaryParentheses()
+            }
+        }
+
+        val parentExclPrefixExpression =
+            callOrQualified.parents.dropWhile { it is KtParenthesizedExpression }.firstOrNull()?.asExclPrefixExpression()
+        psiFactory.buildExpression {
+            appendFixedText(if (negateCall && parentExclPrefixExpression == null) "!" else "")
+            if (callOrQualified is KtQualifiedExpression) {
+                appendExpression(callOrQualified.receiverExpression)
+                appendFixedText(".")
+            }
+            appendFixedText(toFunctionName)
+            element.valueArgumentList?.let {
+                appendFixedText(it.text)
+            }
+            if (element.lambdaArguments.isNotEmpty()) {
+                appendFixedText(lambda.text)
+            }
```

Comment:
It'd be better to extract it to a function to avoid the code duplication from `ConvertToIndexedFunctionCallIntention`

Reply:
955365e